### PR TITLE
U512 ops

### PR DESCRIPTION
**Pull request by zielvna**

### CODE REVIEW

File: contracts/math/uints.ral [hunk 1]
```diff
@@ -90,5 +90,52 @@ Abstract Contract U512Impl () {
...
+        let r3 = a0 * b1
+
+        let mut u512 = bigAdd512(U512 { higher: r1, lower: r0 }, U512 { higher: high256(r2), lower: low256(r2) << 128 })
+        u512 = bigAdd512(u512, U512 { higher: high256(r3), lower: low256(r3) << 128 })
+
+        return u512
+    }
+    
+    pub fn overflowingAdd256(a: U256, b: U256) -> (U256, U256) {
+        if (u256Max!() - a >= b) {
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:118]: There is defined const value for `u256Max!()`

File: contracts/math/uints.ral [hunk 2]
```diff
@@ -90,5 +90,52 @@ Abstract Contract U512Impl () {
...
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd256(a.lower, b.lower)
+        let r1 = wrappingAdd256(wrappingAdd256(a.higher, b.higher), overflow)
+        return U512 { higher: r1, lower: r0 }
+    }
+
+    pub fn bigMul256(a: U256, b: U256) -> U512 {
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:100]: will `bigMulUp` will be added within this PR?

Comment (Sniezka1927) [reply contracts/math/uints.ral:100]: Have you tried to optimalize it? GPT could be usefull

**Review by Sniezka1927 — CHANGES_REQUESTED** [review]

File: contracts/math/uints.ral [hunk 3]
```diff
@@ -90,5 +90,52 @@ Abstract Contract U512Impl () {
...
+        } else {
+            return b + (u256Max!() - a) - 1
+        }
+    }
+
+    pub fn low256(a: U256) -> U256 {
+        return a << 128 >> 128
+    }
+    
+    pub fn high256(a: U256) -> U256 {
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:137]: I suggest to change naming to `high128` and `low128`

File: contracts/math/uints.ral (hunk 1 above)
Comment (wojciech-cichocki) [reply contracts/math/uints.ral:118]: +1

File: contracts/math/uints.ral (hunk 2 above)
Comment (wojciech-cichocki) [reply contracts/math/uints.ral:100]: +1

File: contracts/math/uints.ral [hunk 4]
```diff
@@ -3,7 +3,7 @@ struct U512 {
     mut lower: U256
 }
 
-Abstract Contract U512Impl () {
+Contract Uints () {
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:6]: Please explain the benefits of using a Contract instead of an Abstract Contract

File: contracts/math/uints.ral [hunk 5]
```diff
@@ -90,5 +90,52 @@ Abstract Contract U512Impl () {
...
+    pub fn wrappingAdd256(a: U256, b: U256) -> U256 {
+        if (u256Max!() - a >= b) {
+            return a + b
+        } else {
+            return b + (u256Max!() - a) - 1
+        }
+    }
+
+    pub fn low256(a: U256) -> U256 {
+        return a << 128 >> 128
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:134]: Is it safe to overflow here? I mean, what will happen if I pass the max U256? Initially, the value will be increased by 2^128, which exceeds U256::max. Depending on how these bitwise operators are implemented, they can either trigger or ignore overflow. It's worth testing this case

Comment (wojciech-cichocki) [reply contracts/math/uints.ral:134]: However, IMO extracting the lower u128 can be performed more easily using a bitwise mask: `u128::max & a`

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

File: contracts/math/uints.ral (hunk 4 above)
Comment (zielvna) [reply contracts/math/uints.ral:6]: abstract contract is used to extend normal contract with functions of abstract contract, it will still produce one contract as an output, bytecode limit of a decimal contract (one where we use uints) was exceeded so i had to split it and make uints as its own contract

File: contracts/math/uints.ral (hunk 5 above)
Comment (zielvna) [reply contracts/math/uints.ral:134]: it is safe, it will cut first half of the bits and replace it will zeros, ok ill use different approach

File: test/uints.test.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    }
+    {
+      const a = { higher: 0n, lower: 115792089237316195423570985008687907853269984665640564039457584007913129639935n }
+      const b = { higher: 0n, lower: 20n }
+      const result = (await uints.contractInstance.methods.bigAdd512({ args: { a, b } })).returns
+      expect(result).toStrictEqual({ higher: 1n, lower: 19n })
+    }
+  })
+
+  test('big mul 256', async () => {
```
Comment (wojciech-cichocki) [inline test/uints.test.ts:181]: Missing edge cases:
- mul by `0` (try `a`/`b`)

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

File: contracts/math/uints.ral [hunk 7]
```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
...
+        let mut result = bigAdd512(U512 { higher: rHigherHigher, lower: rLowerLower }, U512 { higher: high128(rHigherLower), lower: low128(rHigherLower) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(rLowerHigher), lower: low128(rLowerHigher) << 128 })
+
+        result = bigDiv(result, denominator, 1)
+
+        return result
+    }
+
+    pub fn bigMulUp(a: U256, b: U256, denominator: U256) -> U512 {
+        let a0 = low128(a)
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:120]: https://github.com/invariant-labs/protocol-alephium/pull/39#discussion_r1547829186

File: contracts/math/uints.ral [hunk 8]
```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
...
+        let aHigher = high128(a)
+        let bLower = low128(b)
+        let bHigher = high128(b)
+
+        let rLowerLower = aLower * bLower
+        let rHigherHigher = aHigher * bHigher
+        let rHigherLower = aHigher * bLower
+        let rLowerHigher = aLower * bHigher
+
+        let mut result = bigAdd512(U512 { higher: rHigherHigher, lower: rLowerLower }, U512 { higher: high128(rHigherLower), lower: low128(rHigherLower) << 128 })
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:111]: I would adjust the naming `rHigherHigher` `rLowerLower` does not seem to be the best option

File: test/uints.test.ts [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    }
+    {
+      const a = { higher: 0n, lower: 115792089237316195423570985008687907853269984665640564039457584007913129639935n }
+      const b = { higher: 0n, lower: 20n }
+      const result = (await uints.contractInstance.methods.bigAdd512({ args: { a, b } })).returns
+      expect(result).toStrictEqual({ higher: 1n, lower: 19n })
+    }
+  })
+
+  test('big mul', async () => {
```
Comment (Sniezka1927) [inline test/uints.test.ts:181]: I would add tests also for big mul up

**Review by Sniezka1927 — CHANGES_REQUESTED** [review]

File: contracts/math/uints.ral [hunk 10]
```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
...
+    pub fn overflowingAdd(a: U256, b: U256) -> (U256, U256) {
+        if (u256Max!() - a >= b) {
+            return a + b, 0
+        } else {
+            return b + (u256Max!() - a) - 1, 1
+        }
+    }
+
+    pub fn wrappingAdd(a: U256, b: U256) -> U256 {
+        let u256Max = u256Max!()
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:148]: You do not need to define additional variable.
It is already defined
`const MaxU256 = 115792089237316195423570985008687907853269984665640564039457584007913129639935`

File: contracts/math/uints.ral [hunk 11]
```diff
@@ -90,5 +90,76 @@ Abstract Contract U512Impl () {
...
+        let r0 = a0 * b0
+        let r1 = a1 * b1
+        let r2 = a1 * b0
+        let r3 = a0 * b1
+
+        let mut result = bigAdd512(U512 { higher: r1, lower: r0 }, U512 { higher: high128(r2), lower: low128(r2) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(r3), lower: low128(r3) << 128 })
+
+        result = bigAdd512(result, U512 { higher: 0, lower: 10 ** denominator - 1 })
+        result = bigDiv(result, denominator, 1)
```
Comment (Sniezka1927) [inline contracts/math/uints.ral:134]: The `denominator` is the scale or is it already calculated?

**Review by Sniezka1927 — APPROVED** [review]

File: contracts/math/uints.ral [hunk 12]
```diff
@@ -90,5 +90,74 @@ Abstract Contract U512Impl () {
                 lower: deltaL
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd(a.lower, b.lower)
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:95]: `r0`, `r1` also here

File: contracts/math/uints.ral [hunk 13]
```diff
@@ -90,5 +90,74 @@ Abstract Contract U512Impl () {
...
+
+        let mut result = bigAdd512(U512 { higher: aHigherBHigher, lower: aLowerBLower }, U512 { higher: high128(aHigherBLower), lower: low128(aHigherBLower) << 128 })
+        result = bigAdd512(result, U512 { higher: high128(aLowerBHigher), lower: low128(aLowerBHigher) << 128 })
+
+        result = bigDiv(result, denominator, 1)
+
+        return result
+    }
+
+    pub fn bigMulUp(a: U256, b: U256, denominator: U256) -> U512 {
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:119]: substantial portion of the code is shared between `bigMul` and `bigMulUp`. As is often the case, numerous points of definition can lead to unexpected inconsistencies if we only replace one part when making changes.

It's worth abstracting this common logic into a separate function

File: contracts/math/uints.ral [hunk 14]
```diff
@@ -90,5 +90,74 @@ Abstract Contract U512Impl () {
...
             }
         }
+
+    pub fn bigAdd512(a: U512, b: U512) -> U512 {
+        let (r0, overflow) = overflowingAdd(a.lower, b.lower)
+        let r1 = wrappingAdd(wrappingAdd(a.higher, b.higher), overflow)
+        return U512 { higher: r1, lower: r0 }
+    }
+
+    pub fn bigMul(a: U256, b: U256, denominator: U256) -> U512 {
```
Comment (wojciech-cichocki) [inline contracts/math/uints.ral:100]: We have a convention that the result of multiplication `a * b` is returned includes `a` denominator. However, this is just convention, and the result could be returned with the denominator `b`.

That's why IMO we should change the name `denominator` to `bDenominator`. Please make this change in the remaining places as well.

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

**Review by wojciech-cichocki — APPROVED** [review]
